### PR TITLE
Tutorial: ESC managed chat app

### PR DESCRIPTION
**Pull request by thoward**
This tutorial covers ESC CLI and ESC SDK showing how to bring a basic Python ChatGPT clone app under management by Pulumi ESC. It shows both static configuration and dynamic configuration using the ESC SDK.

### CODE REVIEW

File: content/tutorials/esc-managed-chatapp/index.md [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+pip install flask openai
+```
+
+### Step 3: Create the webapp
+
+Finally, lets create the code files for our chat app. We'll need two files `chatapp.py`, the backend Flask app, and `templates/chatapp.html`, an HTML template that will render the UI. You can copy and paste these from here, into your favorite code editor and save.
+
+**Code Listing:** `chatapp.py`
+
+```python
```
Comment (interurban) [inline content/tutorials/esc-managed-chatapp/index.md:134]: might consider adding something to give an error if they don't correctly get their API key in: 

import os
from flask import Flask, render_template, request, jsonify
from openai import OpenAI

app = Flask(__name__)

def get_openai_response(api_key, model, prompt):
    client = OpenAI(api_key=api_key)
    try:
        response = client.chat.completions.create(
            model=model,
            messages=[{"role": "user", "content": prompt}]
        )
        # Print the full response for debugging
        print(response)
        # Return only the content of the message
        return response.choices[0].message.content
    except Exception as e:
        return f"An error occurred: {str(e)}"

@app.route('/')
def home():
    return render_template('chatapp.html')

@app.route('/chat', methods=['POST'])
def chat():
    model = "gpt-3.5-turbo"
    api_key = os.getenv("OPENAI_API_KEY")

    prompt = request.json['message']
    response = get_openai_response(api_key, model, prompt)
    return jsonify({'response': response, 'model': model})

if __name__ == '__main__':
    app.run(debug=True)

File: content/tutorials/esc-managed-chatapp/index.md [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+    return jsonify({'response': response, 'model': model})
+
+if __name__ == '__main__':
+    app.run(debug=True)
+
+```
+
+**Code Listing:** `templates/chatapp.html`
+
+```html
```
Comment (interurban) [inline content/tutorials/esc-managed-chatapp/index.md:172]: Just an idea on Pulumi branding :) ; does add 10 or so lines of code, so take it or leave it: 

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>GPT-4 Chat App Using SDK</title>
    <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
    <style>
        body { font-family: Arial, sans-serif; max-width: 800px; margin: 0 auto; padding: 20px; background-color: #5D2E8C; color: white; }
        #chat-container { border: 2px solid #F8633A; height: 400px; overflow-y: scroll; padding: 10px; margin-bottom: 10px; background-color: #ffffff; color: black; }
        #user-input { width: 70%; padding: 5px; border: 2px solid #F8633A; }
        #send-button { padding: 5px 10px; background-color: #F8633A; color: white; border: none; cursor: pointer; }
        #send-button:hover { background-color: #FF8556; }
        #status { color: grey; font-style: italic; margin-bottom: 10px; }
        footer { text-align: center; margin-top: 20px; font-size: 0.9em; color: white; }
    </style>
</head>
<body>
    <h1>OpenAI Chat App Using SDK</h1>
    <div id="chat-container">
        <p id="status">Awaiting your message...</p>
    </div>
    <input type="text" id="user-input" placeholder="Type your message...">
    <button id="send-button">Send</button>

    <footer>
        <p>Made with 🤖 by Pulumi</p>
    </footer>

    <script>
        $(document).ready(function() {
            function sendMessage() {
                var userMessage = $('#user-input').val().trim();
                if (userMessage === '') return;

                $('#status').remove();
                $('#chat-container').append('<p><strong>You:</strong> ' + userMessage + '</p>');
                $('#user-input').val('');
                $('#send-button').prop('disabled', true);

                $.ajax({
                    url: '/chat',
                    method: 'POST',
                    contentType: 'application/json',
                    data: JSON.stringify({message: userMessage}),
                    success: function(response) {
                        $('#chat-container').append('<p><strong>LLM ('+ response.model + '):</strong> ' + response.response + '</p>');
                        $('#chat-container').scrollTop($('#chat-container')[0].scrollHeight);
                    },
                    error: function() {
                        $('#chat-container').append('<p style="color: red;"><strong>Error:</strong> Unable to connect to the chatbot.</p>');
                    },
                    complete: function() {
                        $('#send-button').prop('disabled', false);
                    }
                });
            }

            $('#send-button').click(sendMessage);
            $('#user-input').keypress(function(e) {
                if (e.which == 13) sendMessage();
            });
        });
    </script>
</body>
</html>

File: content/tutorials/esc-managed-chatapp/index.md [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+
+    prompt = request.json['message']
+    response, model = get_openai_response(api_key, model, prompt)
+    return jsonify({'response': response, 'model': model})
+```
+
+Now let's run the app, with ESC providing the variables, to make sure it works.
+
+```bash
+$ esc env run default/chatapp -- python chatapp.py`
```
Comment (interurban) [inline content/tutorials/esc-managed-chatapp/index.md:365]:
```suggestion
$ esc env run default/chatapp -- python chatapp.py
```

File: content/tutorials/esc-managed-chatapp/index.md [hunk 4]
```diff
@@ -0,0 +1,533 @@
...
+```
+
+If this returns nothing, you'll need to [create an access token](/docs/pulumi-cloud/access-management/access-tokens/):
+
+{{< figure src="pulumi-cloud-create-access-token.png" caption="Figure: Create an access token in Pulumi Cloud">}}
+
+Copy that value to the clipboard, and paste it securely into your shell:
+
+```bash
+$ PULUMI_ACCESS_TOKEN=`pbpaste`
```
Comment (interurban) [inline content/tutorials/esc-managed-chatapp/index.md:435]: forgot an export?

Comment (thoward) [reply content/tutorials/esc-managed-chatapp/index.md:435]: indeed

File: content/tutorials/esc-managed-chatapp/index.md (hunk 2 above)
Comment (thoward) [reply content/tutorials/esc-managed-chatapp/index.md:172]: will come back to these suggestions when i have time to apply and retest code.

**Review by interurban — APPROVED** [review]
thx Troy

**Pull request merged**